cmd/agent/app/throttling/credit_store.go

```diff
@@ -0,0 +1,261 @@
+	}
+}
+
+// Spend depletes the balance from the client for the given operation. The usage
+// would be when a client submits spans for operation X, it is spending those
+// credits it requested for operation X.
+func (c *client) Spend(operationName string, credits float64) bool {
+	balance := c.perOperationBalance[operationName]
+	if credits > balance {
+		return false
+	}
+
+	balance -= credits
+	c.perOperationBalance[operationName] = balance
+	return true
+}
+
+// CreditAccruerOptions provides values to be used with a creditAccruer object
+type CreditAccruerOptions struct {
+	MaxOperations    int
+	CreditsPerSecond float64
+	MaxBalance       float64
+	ClientMaxBalance float64
+}
+
+type creditAccruer struct {
+	options                 CreditAccruerOptions
+	serviceName             string
+	perOperationRateLimiter OverrideMap
+	updateTime              time.Time
+	timeNow                 func() time.Time // For testing
+}
+
+func newCreditAccruer(serviceName string, options CreditAccruerOptions, timeNow func() time.Time) *creditAccruer {
+	defaultRateLimiter := newTokenBucket(options, timeNow)
+	perOperationRateLimiter := NewOverrideMap(options.MaxOperations, defaultRateLimiter)
+	return &creditAccruer{
+		options:                 options,
+		serviceName:             serviceName,
+		updateTime:              timeNow(),
+		timeNow:                 timeNow,
+		perOperationRateLimiter: perOperationRateLimiter,
+	}
+}
+
+// Withdraw credits from the creditAccruer for the given operationName with an
+// upper limit of maxWithdrawal.
+func (accruer *creditAccruer) Withdraw(operationName string, maxWithdrawal float64) float64 {
+	var rateLimiter *tokenBucket
+	if accruer.perOperationRateLimiter.Has(operationName) ||
+		accruer.perOperationRateLimiter.IsFull() {
+		rateLimiter = accruer.perOperationRateLimiter.Get(operationName).(*tokenBucket)
+	} else {
+		rateLimiter = newTokenBucket(accruer.options, accruer.timeNow)
+		accruer.perOperationRateLimiter.Set(operationName, rateLimiter)
+	}
+	credits := rateLimiter.Withdraw(maxWithdrawal)
+	return credits
+}
+
+// CreditStore manages the relationship between service instance clients and
+// backend storage credits. Each client must withdraw credits in order to submit
+// trace data to the Jaeger agent. These credits are generated on behalf of the
+// service, of which there can be many instances, and thus many clients.
+type CreditStore interface {
+	io.Closer
+	// Withdraw deducts as many credits from the service on behalf of the client
+	// without exceeding MaxClientBalance. The client is identified by a unique
+	// clientID string. The credits are deducted from the specific operation on
+	// the service if present, otherwise from the default credit pool.
+	Withdraw(clientID string, serviceName string, operationName string) float64
+	// Spend credits already allocated to this client on a given operation.
+	Spend(clientID string, operationName string, credits float64) bool
```

Comment:
need to rethink Withdraw and Spend terminology, the pair doesn't quite make sense.

Conceptually the first operation the client does is to reserve a certain amount of credits, such that they are not available to other clients, but not yet spent by this client. Then it spends from the reserved amount. So maybe Reserve/Spend.

The lack of serviceName in the Spend function looks a bit odd. Maybe clientID should be the second argument in the first method, because service->client  is 1-*.

Reply:
Originally, I had done `Withdraw` and `Deposit`, but realized `Deposit` would imply it went back to the service credit pool. Therefore, switched `Deposit` to `Spend`, but `Reserve`/`Spend` or `Acquire`/`Release` might make more sense.

Regarding the second issue, I even considered adding a `serviceName` argument that would remain unused until I found a use for it, just because the consistency bothers me here too. I'll try your suggestion to see if it looks cleaner.